### DISCUSSION TITLE
How to tell tauri is online or offline ?

### DISCUSSION THREAD

**Discussion by gato25** [tauri-apps/tauri#9269] [Q&A]
Hello, I want to connect remote database or embedded database on connection status and i have tried to use navigator.onLine property and it just returns true even though my internet is disconnected.
How do i solve this ?

**Comment by Bharath-ilangovan** [comment]
> Hello, I want to connect remote database or embedded database on connection status and i have tried to use navigator.onLine property and it just returns true even though my internet is disconnected. How do i solve this ?

 same issues for me too have you got any solution

**Comment by FabianLars** [comment]
navigator.onLine is in my opinion generally not suitable to do online/offline checks, see https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine

The easy workaround is to just try doing a dummy http request (or a bare tcp connection if we're talking about rust) and see if it fails. But even that may not be 100% reliable though :/

**Reply by SultanAlzoghaibi** [reply]
this would require a while loop to check periodically

**Comment by Joaquin121121** [comment]
Im using a custom hook that pings a reliable endpoint 'https://www.google.com/generate_204' every 5 seconds to update connection status

**Comment by vincere-mori** [comment]
Yeah `navigator.onLine` is basically useless for this. It only tells you whether the OS thinks any network interface is up, not whether you actually have internet. VPN, captive portal, hotspot with no upstream, WSL bridge, all of them make it lie.

What I do in my Tauri 2 app is a small Rust command that does a real probe, called on a timer from the frontend. Cheapest reliable check is hitting one of the standard captive-portal endpoints, they return 204 if reachable and the big vendors host them so they're not going anywhere:

```rust
// src-tauri/src/lib.rs
#[tauri::command]
async fn is_online() -> bool {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(3))
        .build()
        .unwrap();
    client
        .get("https://www.gstatic.com/generate_204")
        .send()
        .await
        .map(|r| r.status().as_u16() == 204)
        .unwrap_or(false)
}
```

Register it:
```rust
.invoke_handler(tauri::generate_handler![is_online])
```

In `Cargo.toml`:
```toml
reqwest = { version = "0.12", default-features = false, features = ["rustls-tls"] }
```
rustls keeps the binary small and avoids the OpenSSL headache on Windows builds.

Frontend side:
```ts
import { invoke } from "@tauri-apps/api/core";

export function watchConnection(onChange: (online: boolean) => void) {
  let last: boolean | null = null;
  return setInterval(async () => {
    const now = await invoke<boolean>("is_online");
    if (now !== last) { last = now; onChange(now); }
  }, 5000);
}
```

One thing worth keeping in mind for your remote-vs-embedded DB switch: don't probe your own backend for this. If your API is down it'll look like the user is offline and you'll fall back to embedded DB even when the real internet is fine. Also if you care about usable connectivity, treat the probe as failed when it takes longer than ~2s, a 30s timeout connection is functionally offline for most UX.
